{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Requirements\n",
    "For x64 Linux or macOS:\n",
    "```\n",
    "conda create -n orthofinder orthofinder\n",
    "```\n",
    "\n",
    "For arm64 macOS using Rosetta:\n",
    "```\n",
    "conda create -n orthofinder\n",
    "conda activate orthofinder\n",
    "conda config --env --set subdir osx-64\n",
    "conda install orthofinder\n",
    "```\n",
    "\n",
    "Then set kernal of this Jupter Notebook to Python in orthofinder Conda environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Astatotilapia_calliptera.fAstCal1.3.pep.all.fa.gz']\n",
      "['Maylandia_zebra.M_zebra_UMD2a.pep.all.fa.gz']\n",
      "['Oreochromis_niloticus.O_niloticus_UMD_NMBU.pep.all.fa.gz']\n",
      "['Danio_rerio.GRCz11.pep.all.fa.gz']\n",
      "['Amphilophus_citrinellus.Midas_v5.pep.all.fa.gz']\n",
      "['Pundamilia_nyererei.PunNye1.0.pep.all.fa.gz']\n",
      "['Haplochromis_burtoni.AstBur1.0.pep.all.fa.gz']\n",
      "['Neolamprologus_brichardi.NeoBri1.0.pep.all.fa.gz']\n",
      "['Oryzias_latipes.ASM223467v1.pep.all.fa.gz']\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import re\n",
    "import requests\n",
    "import subprocess\n",
    "\n",
    "species_list = [\"Astatotilapia calliptera\", \"Maylandia zebra\", \"Oreochromis niloticus\", \"Danio rerio\", \"Amphilophus citrinellus\", \"Pundamilia nyererei\", \"Haplochromis burtoni\", \"Neolamprologus brichardi\", \"Oryzias latipes\"]\n",
    "\n",
    "os.makedirs('pep', exist_ok=True)\n",
    "for species in species_list:\n",
    "    species = species.replace(' ', '_').lower()\n",
    "    response = requests.get(f\"https://ftp.ensembl.org/pub/current_fasta/{species}/pep/\")\n",
    "    genome = re.findall(r'href=\"([^\"]+\\.pep\\.all\\.fa\\.gz)\"', response.text)\n",
    "    print(genome)\n",
    "    if genome:\n",
    "        genome_file = genome[0]\n",
    "        genome_path = os.path.join('pep/original', genome_file)\n",
    "        unzip_path = genome_path.replace('.gz', '')\n",
    "        if not os.path.isfile(genome_path) and not os.path.isfile(unzip_path):\n",
    "            dna_url = f\"https://ftp.ensembl.org/pub/current_fasta/{species}/pep/{genome_file}\"\n",
    "            response = requests.get(dna_url)\n",
    "            with open(genome_path, 'wb') as f:\n",
    "                f.write(response.content)\n",
    "        if not os.path.isfile(unzip_path):\n",
    "            subprocess.run(['gunzip', '-k', genome_path])\n",
    "        os.remove(genome_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import requests\n",
    "import subprocess\n",
    "\n",
    "primary_transcript_script = 'pep/primary_transcript.py'\n",
    "if not os.path.isfile(primary_transcript_script):\n",
    "    url = \"https://raw.githubusercontent.com/davidemms/OrthoFinder/refs/heads/master/tools/primary_transcript.py\"\n",
    "    response = requests.get(url)\n",
    "    with open(primary_transcript_script, 'wb') as f:\n",
    "        f.write(response.content)\n",
    "if not os.path.exists('pep/original/primary_transcripts'):\n",
    "    fa_files = [f for f in os.listdir('pep/original') if f.endswith('.fa')]\n",
    "    for f in fa_files:\n",
    "        subprocess.run(['python', primary_transcript_script, os.path.join('pep/original', f)])\n",
    "for file in os.listdir('pep/original/primary_transcripts'):\n",
    "    os.rename(os.path.join('pep/original/primary_transcripts', file), os.path.join('pep', file))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "zebrafish_file = 'pep/original/Danio_rerio.GRCz11.pep.all.fa'\n",
    "filtered_file = 'pep/Danio_rerio.alt_ids.txt'\n",
    "\n",
    "with open(zebrafish_file, 'r') as infile, open(filtered_file, 'w') as outfile:\n",
    "    write = True\n",
    "    for line in infile:\n",
    "        if line.startswith('>'):\n",
    "            if 'ALT_' in line:\n",
    "                gene_id = line.split('gene:')[1].split()[0]\n",
    "                outfile.write(f'{gene_id}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OrthoFinder version 2.5.5 Copyright (C) 2014 David Emms\n",
      "\n",
      "2024-11-21 13:34:13 : Starting OrthoFinder 2.5.5\n",
      "14 thread(s) for highly parallel tasks (BLAST searches etc.)\n",
      "1 thread(s) for OrthoFinder algorithm\n",
      "\n",
      "Checking required programs are installed\n",
      "----------------------------------------\n",
      "Test can run \"mcl -h\" - ok\n",
      "Test can run \"fastme -i /Users/home/Downloads/Project-Cichlids/pep/OrthoFinder/Results_Nov21/WorkingDirectory/dependencies/SimpleTest.phy -o /Users/home/Downloads/Project-Cichlids/pep/OrthoFinder/Results_Nov21/WorkingDirectory/dependencies/SimpleTest.tre\" - ok\n",
      "\n",
      "WARNING: Files have been ignored as they don't appear to be FASTA files:\n",
      ".DS_Store\n",
      "primary_transcript.py\n",
      "OrthoFinder expects FASTA files to have one of the following extensions: fasta, fa, faa, pep, fas\n",
      "\n",
      "Dividing up work for BLAST for parallel processing\n",
      "--------------------------------------------------\n",
      "2024-11-21 13:34:15 : Creating diamond database 1 of 9\n",
      "2024-11-21 13:34:15 : Creating diamond database 2 of 9\n",
      "2024-11-21 13:34:15 : Creating diamond database 3 of 9\n",
      "2024-11-21 13:34:16 : Creating diamond database 4 of 9\n",
      "2024-11-21 13:34:16 : Creating diamond database 5 of 9\n",
      "2024-11-21 13:34:16 : Creating diamond database 6 of 9\n",
      "2024-11-21 13:34:16 : Creating diamond database 7 of 9\n",
      "2024-11-21 13:34:16 : Creating diamond database 8 of 9\n",
      "2024-11-21 13:34:16 : Creating diamond database 9 of 9\n",
      "\n",
      "Running diamond all-versus-all\n",
      "------------------------------\n",
      "Using 14 thread(s)\n",
      "2024-11-21 13:34:16 : This may take some time....\n",
      "2024-11-21 13:34:16 : Done 0 of 81\n",
      "2024-11-21 13:39:55 : Done 10 of 81\n",
      "2024-11-21 13:43:48 : Done 20 of 81\n",
      "2024-11-21 13:47:34 : Done 30 of 81\n",
      "2024-11-21 13:50:02 : Done 40 of 81\n",
      "2024-11-21 13:53:32 : Done 50 of 81\n",
      "2024-11-21 13:56:39 : Done 60 of 81\n",
      "2024-11-21 14:02:43 : Done all-versus-all sequence search\n",
      "\n",
      "Running OrthoFinder algorithm\n",
      "-----------------------------\n",
      "2024-11-21 14:02:44 : Initial processing of each species\n",
      "2024-11-21 14:02:57 : Initial processing of species 0 complete\n",
      "2024-11-21 14:03:13 : Initial processing of species 1 complete\n",
      "2024-11-21 14:03:31 : Initial processing of species 2 complete\n",
      "2024-11-21 14:03:44 : Initial processing of species 3 complete\n",
      "2024-11-21 14:04:01 : Initial processing of species 4 complete\n",
      "2024-11-21 14:04:14 : Initial processing of species 5 complete\n",
      "2024-11-21 14:04:32 : Initial processing of species 6 complete\n",
      "2024-11-21 14:04:46 : Initial processing of species 7 complete\n",
      "2024-11-21 14:04:59 : Initial processing of species 8 complete\n",
      "2024-11-21 14:05:11 : Connected putative homologues\n",
      "2024-11-21 14:05:13 : Written final scores for species 0 to graph file\n",
      "2024-11-21 14:05:15 : Written final scores for species 1 to graph file\n",
      "2024-11-21 14:05:17 : Written final scores for species 2 to graph file\n",
      "2024-11-21 14:05:18 : Written final scores for species 3 to graph file\n",
      "2024-11-21 14:05:20 : Written final scores for species 4 to graph file\n",
      "2024-11-21 14:05:22 : Written final scores for species 5 to graph file\n",
      "2024-11-21 14:05:24 : Written final scores for species 6 to graph file\n",
      "2024-11-21 14:05:25 : Written final scores for species 7 to graph file\n",
      "2024-11-21 14:05:27 : Written final scores for species 8 to graph file\n",
      "2024-11-21 14:05:53 : Ran MCL\n",
      "\n",
      "Writing orthogroups to file\n",
      "---------------------------\n",
      "OrthoFinder assigned 223252 genes (97.5% of total) to 22185 orthogroups. Fifty percent of all genes were in orthogroups with 9 or more genes (G50 was 9) and were contained in the largest 7388 orthogroups (O50 was 7388). There were 13195 orthogroups with all species present and 7587 of these consisted entirely of single-copy genes.\n",
      "\n",
      "2024-11-21 14:05:59 : Done orthogroups\n",
      "\n",
      "Analysing Orthogroups\n",
      "=====================\n",
      "\n",
      "Calculating gene distances\n",
      "--------------------------\n",
      "2024-11-21 14:07:59 : Done\n",
      "\n",
      "Inferring gene and species trees\n",
      "--------------------------------\n",
      "2024-11-21 14:08:01 : Done 0 of 20136\n",
      "2024-11-21 14:08:05 : Done 1000 of 20136\n",
      "2024-11-21 14:08:09 : Done 2000 of 20136\n",
      "2024-11-21 14:08:13 : Done 3000 of 20136\n",
      "2024-11-21 14:08:17 : Done 4000 of 20136\n",
      "2024-11-21 14:08:21 : Done 5000 of 20136\n",
      "2024-11-21 14:08:25 : Done 6000 of 20136\n",
      "2024-11-21 14:08:29 : Done 7000 of 20136\n",
      "2024-11-21 14:08:33 : Done 8000 of 20136\n",
      "2024-11-21 14:08:37 : Done 9000 of 20136\n",
      "2024-11-21 14:08:41 : Done 10000 of 20136\n",
      "2024-11-21 14:08:44 : Done 11000 of 20136\n",
      "2024-11-21 14:08:48 : Done 12000 of 20136\n",
      "2024-11-21 14:08:52 : Done 13000 of 20136\n",
      "2024-11-21 14:08:56 : Done 14000 of 20136\n",
      "2024-11-21 14:09:00 : Done 15000 of 20136\n",
      "2024-11-21 14:09:04 : Done 16000 of 20136\n",
      "2024-11-21 14:09:08 : Done 17000 of 20136\n",
      "2024-11-21 14:09:12 : Done 18000 of 20136\n",
      "2024-11-21 14:09:16 : Done 19000 of 20136\n",
      "2024-11-21 14:09:20 : Done 20000 of 20136\n",
      "\n",
      "13195 trees had all species present and will be used by STAG to infer the species tree\n",
      "\n",
      "Best outgroup(s) for species tree\n",
      "---------------------------------\n",
      "2024-11-21 14:15:59 : Starting STRIDE\n",
      "2024-11-21 14:16:00 : Done STRIDE\n",
      "Observed 245 well-supported, non-terminal duplications. 241 support the best root and 4 contradict it.\n",
      "Best outgroup for species tree:\n",
      "  Danio_rerio.GRCz11.pep.all\n",
      "\n",
      "Reconciling gene trees and species tree\n",
      "---------------------------------------\n",
      "Outgroup: Danio_rerio.GRCz11.pep.all\n",
      "2024-11-21 14:16:00 : Starting Recon and orthologues\n",
      "2024-11-21 14:16:00 : Starting OF Orthologues\n",
      "2024-11-21 14:16:00 : Done 0 of 20136\n",
      "2024-11-21 14:16:06 : Done 1000 of 20136\n",
      "2024-11-21 14:16:08 : Done 2000 of 20136\n",
      "2024-11-21 14:16:10 : Done 3000 of 20136\n",
      "2024-11-21 14:16:12 : Done 4000 of 20136\n",
      "2024-11-21 14:16:13 : Done 5000 of 20136\n",
      "2024-11-21 14:16:15 : Done 6000 of 20136\n",
      "2024-11-21 14:16:17 : Done 7000 of 20136\n",
      "2024-11-21 14:16:18 : Done 8000 of 20136\n",
      "2024-11-21 14:16:19 : Done 9000 of 20136\n",
      "2024-11-21 14:16:21 : Done 10000 of 20136\n",
      "2024-11-21 14:16:22 : Done 11000 of 20136\n",
      "2024-11-21 14:16:24 : Done 12000 of 20136\n",
      "2024-11-21 14:16:25 : Done 13000 of 20136\n",
      "2024-11-21 14:16:26 : Done 14000 of 20136\n",
      "2024-11-21 14:16:28 : Done 15000 of 20136\n",
      "2024-11-21 14:16:29 : Done 16000 of 20136\n",
      "2024-11-21 14:16:30 : Done 17000 of 20136\n",
      "2024-11-21 14:16:31 : Done 18000 of 20136\n",
      "2024-11-21 14:16:33 : Done 19000 of 20136\n",
      "2024-11-21 14:16:33 : Done 20000 of 20136\n",
      "2024-11-21 14:16:33 : Done OF Orthologues\n",
      "\n",
      "Writing results files\n",
      "=====================\n",
      "2024-11-21 14:16:37 : Done orthologues\n",
      "\n",
      "Results:\n",
      "    /Users/home/Downloads/Project-Cichlids/pep/OrthoFinder/Results_Nov21/\n",
      "\n",
      "CITATION:\n",
      " When publishing work that uses OrthoFinder please cite:\n",
      " Emms D.M. & Kelly S. (2019), Genome Biology 20:238\n",
      "\n",
      " If you use the species tree in your work then please also cite:\n",
      " Emms D.M. & Kelly S. (2017), MBE 34(12): 3267-3278\n",
      " Emms D.M. & Kelly S. (2018), bioRxiv https://doi.org/10.1101/267914\n"
     ]
    }
   ],
   "source": [
    "if not os.path.exists('pep/OrthoFinder'):\n",
    "    subprocess.run([\"orthofinder\", \"-f\", \"pep\"])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "osx64",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
