{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "raw"
    }
   },
   "source": [
    "### Requirements\n",
    "`conda create -n <ENV_NAME> --file requirements.txt`\n",
    "\n",
    "Then set kernal of this Jupter Notebook to Python in <ENV_NAME> Conda environment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### OrthoFinder\n",
    "Run OrthoFinder.ipynb first"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "numbl (ENSDARG00000101949) acts upstream of or within GO:0007219\n",
      "maml1 (ENSDARG00000076466) involved in GO:0007219\n",
      "dtx3lb.2 (ENSDARG00000104649) involved in GO:0007219\n",
      "bmp2k (ENSDARG00000020009) involved in GO:0007219\n",
      "adam10a (ENSDARG00000053468) involved in GO:0007219\n",
      "adam17a (ENSDARG00000043213) involved in GO:0007219\n",
      "tet3 (ENSDARG00000062646) acts upstream of or within GO:0007219\n",
      "wnt16 (ENSDARG00000110801) acts upstream of or within GO:0007219\n",
      "dab2 (ENSDARG00000031761) acts upstream of or within GO:0007219\n",
      "prss23 (ENSDARG00000055786) acts upstream of or within GO:0007219\n",
      "zmp:0000001236 (ENSDARG00000112977) involved in GO:0007219\n",
      "susd5 (ENSDARG00000088277) involved in GO:0007219\n",
      "dtx3 (ENSDARG00000025766) involved in GO:0007219\n",
      "aak1b (ENSDARG00000077686) involved in GO:0007219\n",
      "si:ch211-194k22.8 (ENSDARG00000096711) involved in GO:0007219\n",
      "neurl1aa (ENSDARG00000069438) involved in GO:0007219\n",
      "si:dkey-3h3.3 (ENSDARG00000067718) involved in GO:0007219\n",
      "bloc1s2 (ENSDARG00000046081) acts upstream of or within GO:0007219\n",
      "rbpjb (ENSDARG00000052091) involved in GO:0007219\n",
      "wnt16 (ENSDARG00000094973) acts upstream of or within GO:0007219\n",
      "angptl2b (ENSDARG00000068369) acts upstream of or within GO:0007219\n",
      "appb (ENSDARG00000055543) acts upstream of or within GO:0007219\n",
      "hey1 (ENSDARG00000070538) involved in GO:0007219\n",
      "epas1b (ENSDARG00000057671) acts upstream of or within GO:0007219\n",
      "fbxw7 (ENSDARG00000060994) acts upstream of or within GO:0007219\n",
      "jag1a (ENSDARG00000030289) involved in GO:0007219\n",
      "arrb1 (ENSDARG00000043241) acts upstream of or within GO:0007219\n",
      "bbs4 (ENSDARG00000063522) acts upstream of or within GO:0007219\n",
      "ythdf2 (ENSDARG00000014498) acts upstream of or within GO:0007219\n",
      "tet2 (ENSDARG00000076928) acts upstream of or within GO:0007219\n",
      "epas1a (ENSDARG00000008697) acts upstream of or within GO:0007219\n",
      "dtx3la (ENSDARG00000089511) involved in GO:0007219\n",
      "llgl1 (ENSDARG00000009693) acts upstream of or within GO:0007219\n",
      "kif3a (ENSDARG00000087538) acts upstream of GO:0007219\n",
      "pappa2 (ENSDARG00000076020) acts upstream of or within GO:0007219\n",
      "plvapb (ENSDARG00000045003) acts upstream of or within GO:0007219\n",
      "plvapb (ENSDARG00000116150) acts upstream of or within GO:0007219\n",
      "psen1 (ENSDARG00000004870) involved in GO:0007219\n",
      "nle1 (ENSDARG00000057105) involved in GO:0007219\n",
      "hif1aa (ENSDARG00000006181) acts upstream of or within GO:0007219\n",
      "gmds (ENSDARG00000026629) acts upstream of or within GO:0007219\n",
      "pofut1 (ENSDARG00000008953) involved in GO:0007219\n",
      "mettl3 (ENSDARG00000012827) acts upstream of or within GO:0007219\n",
      "dla (ENSDARG00000010791) involved in GO:0007219\n",
      "galnt11 (ENSDARG00000063636) involved in GO:0007219\n",
      "jag1b (ENSDARG00000013168) involved in GO:0007219\n",
      "notch2 (ENSDARG00000043130) involved in GO:0007219\n",
      "mib1 (ENSDARG00000102184) involved in GO:0007219\n",
      "neurod1 (ENSDARG00000019566) acts upstream of or within GO:0007219\n",
      "dlb (ENSDARG00000004232) involved in GO:0007219\n",
      "hif1an (ENSDARG00000031915) involved in GO:0007219\n",
      "crb2a (ENSDARG00000060106) acts upstream of or within GO:0007219\n",
      "yjefn3 (ENSDARG00000061692) involved in GO:0007219\n",
      "fsd1 (ENSDARG00000062017) acts upstream of GO:0007219\n",
      "poglut1 (ENSDARG00000053463) involved in GO:0007219\n",
      "dtx3 (ENSDARG00000116890) involved in GO:0007219\n",
      "rras (ENSDARG00000006553) acts upstream of or within GO:0007219\n",
      "ptbp1a (ENSDARG00000019362) acts upstream of or within GO:0007219\n",
      "uchl3 (ENSDARG00000030177) acts upstream of or within GO:0007219\n",
      "tcima (ENSDARG00000079497) involved in GO:0007219\n",
      "tcima (ENSDARG00000110999) involved in GO:0007219\n",
      "tcima (ENSDARG00000115904) involved in GO:0007219\n",
      "il10 (ENSDARG00000078147) acts upstream of or within GO:0007219\n",
      "arl6 (ENSDARG00000032056) acts upstream of or within GO:0007219\n",
      "arl6 (ENSDARG00000113281) acts upstream of or within GO:0007219\n",
      "angptl1a (ENSDARG00000012071) acts upstream of or within GO:0007219\n",
      "chac1 (ENSDARG00000070426) involved in GO:0007219\n",
      "jag2b (ENSDARG00000021389) involved in GO:0007219\n",
      "mfng (ENSDARG00000042925) involved in GO:0007219\n",
      "uxt (ENSDARG00000019339) acts upstream of or within GO:0007219\n",
      "egfl6 (ENSDARG00000045958) acts upstream of or within GO:0007219\n",
      "hif1ab (ENSDARG00000034293) acts upstream of or within GO:0007219\n",
      "pkd2 (ENSDARG00000014098) acts upstream of GO:0007219\n",
      "chsy1 (ENSDARG00000079027) acts upstream of or within GO:0007219\n",
      "her6 (ENSDARG00000006514) acts upstream of or within GO:0007219\n",
      "nlk1 (ENSDARG00000006566) acts upstream of or within GO:0007219\n",
      "her12 (ENSDARG00000032963) acts upstream of or within GO:0007219\n",
      "asb11 (ENSDARG00000056561) acts upstream of or within GO:0007219\n",
      "llgl2 (ENSDARG00000023920) involved in GO:0007219\n",
      "id2a (ENSDARG00000055283) acts upstream of or within GO:0007219\n",
      "her15.1 (ENSDARG00000054562) acts upstream of or within GO:0007219\n",
      "nrarpb (ENSDARG00000018958) involved in GO:0007219\n",
      "nrarpa (ENSDARG00000091419) involved in GO:0007219\n",
      "nrarpa (ENSDARG00000111485) involved in GO:0007219\n",
      "rab5c (ENSDARG00000026712) acts upstream of or within GO:0007219\n",
      "fezf2 (ENSDARG00000070677) acts upstream of or within GO:0007219\n",
      "heyl (ENSDARG00000055798) involved in GO:0007219\n",
      "heyl (ENSDARG00000112770) involved in GO:0007219\n",
      "aph1b (ENSDARG00000005612) involved in GO:0007219\n",
      "aph1b (ENSDARG00000111794) involved in GO:0007219\n",
      "psenen (ENSDARG00000068698) involved in GO:0007219\n",
      "lfng (ENSDARG00000037879) involved in GO:0007219\n",
      "her4.1 (ENSDARG00000056732) acts upstream of or within GO:0007219\n",
      "psen2 (ENSDARG00000015540) acts upstream of or within GO:0007219\n",
      "foxc1a (ENSDARG00000091481) acts upstream of or within GO:0007219\n",
      "hey2 (ENSDARG00000013441) acts upstream of or within GO:0007219\n",
      "adam17b (ENSDARG00000093093) involved in GO:0007219\n",
      "cad (ENSDARG00000041895) acts upstream of or within GO:0007219\n",
      "notchl (ENSDARG00000088308) involved in GO:0007219\n",
      "dtx3lb.2 (ENSDARG00000075504) involved in GO:0007219\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import glob\n",
    "import pandas as pd\n",
    "import requests\n",
    "\n",
    "base_path = \"pep/OrthoFinder/Results_\"\n",
    "species = \"Astatotilapia_calliptera\"\n",
    "reference_species = \"Danio_rerio\"\n",
    "go_id = \"GO:0007219\" # Notch signaling pathway\n",
    "\n",
    "latest_dir = max(glob.glob(base_path + \"*/\"), key=os.path.getmtime)\n",
    "df_orthogroups = pd.read_csv(os.path.join(latest_dir, \"Orthogroups/Orthogroups.tsv\"), sep='\\t')\n",
    "reference_group = df_orthogroups.filter(like=reference_species).columns[0]\n",
    "species_group = df_orthogroups.filter(like=species).columns[0]\n",
    "\n",
    "def get_taxon_id(species):\n",
    "    url = f\"https://rest.ensembl.org/taxonomy/id/{species}?content-type=application/json\"\n",
    "    response = requests.get(url)\n",
    "    if response.ok:\n",
    "        data = response.json()\n",
    "        return data['id']\n",
    "    else:\n",
    "        response.raise_for_status()\n",
    "\n",
    "def get_genes_from_go(go_id, species):\n",
    "    server = \"https://www.ebi.ac.uk\"\n",
    "    taxon_id = get_taxon_id(species)\n",
    "    genes = []\n",
    "    headers = {\"Content-Type\": \"application/json\"}\n",
    "    i = 1\n",
    "    while True:\n",
    "        ext = f\"/QuickGO/services/annotation/search?goId={go_id}&taxonId={taxon_id}&limit=200&page={i}\"\n",
    "        response = requests.get(server + ext, headers=headers)\n",
    "        if response.status_code >= 400 and response.status_code < 500:\n",
    "            break\n",
    "        if response.ok:\n",
    "            data = response.json()\n",
    "            if not data['results']:\n",
    "                break\n",
    "            for result in data['results']:\n",
    "                url = f\"https://rest.ensembl.org/xrefs/symbol/{species}/{result['geneProductId'].split(':')[1]}?content-type=application/json\"\n",
    "                response = requests.get(url, headers=headers)\n",
    "                if response.ok:\n",
    "                    xrefs = response.json()\n",
    "                    for entry in xrefs:\n",
    "                        if entry['type'] == 'gene':\n",
    "                            if entry['id'] not in [gene['id'] for gene in genes]:\n",
    "                                genes.append({\n",
    "                                    'id': entry['id'],\n",
    "                                    'name': result['symbol'],\n",
    "                                    'qualifier': result['qualifier']\n",
    "                                })\n",
    "                                print(f\"{result['symbol']} ({entry['id']}) {result['qualifier'].replace('_', ' ')} {go_id}\")\n",
    "                else:\n",
    "                    response.raise_for_status()\n",
    "            i += 1\n",
    "        else:\n",
    "            response.raise_for_status()\n",
    "    return genes\n",
    "\n",
    "go_genes = get_genes_from_go(go_id, reference_species)\n",
    "\n",
    "df_go = df_orthogroups[df_orthogroups[reference_group].apply(lambda x: any(gene['id'] in x for gene in go_genes) if pd.notna(x) else False)]\n",
    "df_go.insert(0, 'Matched_Name', df_go[reference_group].apply(lambda x: ', '.join([gene['name'] for gene in go_genes if gene['id'] in x]) if pd.notna(x) else None))\n",
    "df_go.insert(1, 'Matched_ID', df_go[reference_group].apply(lambda x: ', '.join([gene['id'] for gene in go_genes if gene['id'] in x]) if pd.notna(x) else None))\n",
    "df_go.insert(2, 'Matched_Qualifier', df_go[reference_group].apply(lambda x: ', '.join([gene['qualifier'] for gene in go_genes if gene['id'] in x]) if pd.notna(x) else None))\n",
    "reference_counts = df_go[reference_group].apply(lambda x: str(x).count(',') if pd.notna(x) else 0)\n",
    "species_counts = df_go[species_group].apply(lambda x: str(x).count(',') if pd.notna(x) else 0)\n",
    "df_go.insert(3, 'Diff', species_counts - reference_counts)\n",
    "df_go.to_csv(\"go_orthogroups.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Find Differential Exon Usage\n",
    "Run rMATS.ipynb first"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ENSACLG00000016414\n",
    "ENSACLG00000010182\n",
    "/ENSACLG00000010776/\n",
    "ENSACLG00000011505 #Primer design failed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_csv = (\"./dexseq/dexseq_results_filtered_with_norm_counts.csv\", ',')\n",
    "df_csv_sortedby = [\"log2fold_Tropheops_Salima\", 1]\n",
    "df_csv_pval = 'padj'\n",
    "df_csv_groupby = 'groupID'\n",
    "df_csv_rm = [False, 'geneName']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_csv = (\"./rmats/SE.MATS.JC.txt\", '\\t')\n",
    "df_csv_sortedby = [\"IncLevelDifference\", 0.5]\n",
    "df_csv_pval = 'FDR'\n",
    "df_csv_groupby = 'GeneID'\n",
    "df_csv_rm = [False, 'geneSymbol']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_csv = (\"./rmats/MXE.MATS.JCEC.txt\", '\\t')\n",
    "df_csv_sortedby = [\"IncLevelDifference\", 0.5]\n",
    "df_csv_pval = 'FDR'\n",
    "df_csv_groupby = 'GeneID'\n",
    "df_csv_rm = [False, 'geneSymbol']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df_results = pd.read_csv(df_csv[0], delimiter=df_csv[1])\n",
    "df_results_cleaned = df_results[df_results[df_csv_pval] < 0.05].dropna()\n",
    "df_results_cleaned = df_results_cleaned[df_results_cleaned[df_csv_sortedby[0]].abs() >= df_csv_sortedby[1]]\n",
    "df_results_cleaned['max'] = df_results_cleaned.groupby(df_csv_groupby)[df_csv_sortedby[0]].transform(lambda x: x.abs().max())\n",
    "df_results_cleaned = df_results_cleaned.sort_values(by='max', ascending=False).drop(columns=['max'])\n",
    "if df_csv_rm[0]:\n",
    "    df_results_cleaned = df_results_cleaned[df_results_cleaned[df_csv_rm[1]].notna()]\n",
    "df_csv[1][:df_csv[1].rfind('.')]\n",
    "df_results_cleaned.to_csv(f\"{df_csv[0].rsplit('.', 1)[0]}_cleaned.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preparing Sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_id_list = [\"ENSACLG00000010182\", \"ENSACLG00000016414\", \"ENSACLG00000011505\"] # List of genes that might exhibit differential alternative splicing\n",
    "primer_dict = {\"ENSACLG00000010182\": [\"tgcagacggacatccatgag\", \"ccggtggcatactccatctc\", \"agctgcagatggacaacctg\", \"catgaagagaggttccccgg\"]} # Annotate primers that are already designed\n",
    "primer_dict[\"ENSACLG00000016414\"] = [\"CTTCGAGCTGCACGAAATCG\", \"GAAAGCAACGCCTGACTTGG\"] # On the 9th and 9th:10th exon\n",
    "primer_dict[\"ENSACLG00000016414\"] += [\"AGGTGTTTGATGGCGACGAT\", \"TGCTGTCTGAGACGGAAACC\", \"TCGCTGAAGGACGGAGGCGA\"] # On the 8th, 10th, and 9th exons\n",
    "exon_roi = {}\n",
    "#exon_roi = {\"ENSACLG00000016414\": (7, 22)} # Which exons are differentially expressed, this set desgined primer pair 1:2\n",
    "exon_roi = {\"ENSACLG00000016414\": (9, 9)} # For primer pair 3:5\n",
    "#exon_roi = {\"ENSACLG00000011505\": (4, 6)} # Unsucessful"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene: dgkq-ENSACLG00000016414, ROI: Exon9 to 9\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "import requests\n",
    "import re\n",
    "import random\n",
    "import os\n",
    "\n",
    "def get_gene(gene_id):\n",
    "    url = f\"https://rest.ensembl.org/lookup/id/{gene_id}?expand=1\"\n",
    "    headers = {\"Content-Type\": \"application/json\"}\n",
    "    response = requests.get(url, headers=headers)\n",
    "    if not response.ok:\n",
    "        response.raise_for_status()\n",
    "    data = response.json()\n",
    "    return data\n",
    "\n",
    "def get_longest_transcript(gene):\n",
    "    transcripts = gene.get('Transcript', [])\n",
    "    if not transcripts:\n",
    "        raise ValueError(f\"No transcripts found for gene_id: {gene_id}\")\n",
    "    longest_transcript = max(transcripts, key=lambda t: t['length'])\n",
    "    return longest_transcript\n",
    "\n",
    "def get_cds(transcript_id):\n",
    "    url = f\"https://rest.ensembl.org/sequence/id/{transcript_id}?type=cds\"\n",
    "    headers = {\"Content-Type\": \"application/json\"}\n",
    "    response = requests.get(url, headers=headers)\n",
    "    if not response.ok:\n",
    "        response.raise_for_status()\n",
    "    data = response.json()\n",
    "    return data.get('seq', \"\")\n",
    "\n",
    "def get_sequence(seq_id):\n",
    "    url = f\"https://rest.ensembl.org/sequence/id/{seq_id}\"\n",
    "    headers = {\"Content-Type\": \"application/json\"}\n",
    "    response = requests.get(url, headers=headers)\n",
    "    if not response.ok:\n",
    "        response.raise_for_status()\n",
    "    data = response.json()\n",
    "    return data.get('seq', \"\")\n",
    "\n",
    "def random_color():\n",
    "    return \"#{:06x}\".format(random.randint(0, 0xFFFFFF))\n",
    "\n",
    "def annotate_gene(gene_id):\n",
    "    gene = get_gene(gene_id)\n",
    "    gene_description = gene['description'] if 'description' in gene else \"\"\n",
    "    gene_name = gene['display_name'] + \"-\" if 'display_name' in gene else \"\"\n",
    "    if not gene_description or not gene_name:\n",
    "        url = f\"https://rest.ensembl.org/xrefs/id/{gene_id}\"\n",
    "        headers = {\"Content-Type\": \"application/json\"}\n",
    "        response = requests.get(url, headers=headers)\n",
    "        if not response.ok:\n",
    "            response.raise_for_status()\n",
    "        xref_data = response.json()\n",
    "        if xref_data:\n",
    "            gene_name = xref_data[0]['display_id'] + \"-\" if 'display_id' in xref_data[0] else \"\"\n",
    "            gene_description = xref_data[0]['description'] if 'description' in xref_data[0] else \"\"\n",
    "    gene_name = gene_name.lower()\n",
    "    gene_name_list[gene_id] = gene_name\n",
    "    longest_transcript = get_longest_transcript(gene)\n",
    "    longest_transcript_id = longest_transcript['id']\n",
    "    longest_transcript_exons = longest_transcript['Exon']\n",
    "    longest_transcript_cds = get_cds(longest_transcript_id)\n",
    "    cds_seq = Seq(longest_transcript_cds)\n",
    "    cds_record = SeqRecord(cds_seq, id=longest_transcript_id, description=gene_description, annotations={\"molecule_type\": \"DNA\"})\n",
    "    i = 0\n",
    "    if gene_id in exon_roi:\n",
    "        roi_start, roi_end = exon_roi[gene_id]\n",
    "        print(f\"Gene: {gene_name}{gene_id}, ROI: Exon{roi_start} to {roi_end}\")\n",
    "        roi[gene_id] = {'ROI_len':\"\", 'ROI_flank_len':\"\", 'ROI_seq':\"\", 'ROI_flank_seq':\"\"}\n",
    "    strand = longest_transcript['strand']\n",
    "    for exon in longest_transcript_exons:\n",
    "        if strand == 1:\n",
    "            if i == 0:\n",
    "                exon_start = 0\n",
    "                exon_end = exon['end'] - longest_transcript['Translation']['start'] + 1\n",
    "            else:\n",
    "                exon_start = exon_end_last\n",
    "                exon_end = exon_start + exon['end'] - exon['start'] + 1\n",
    "        else:\n",
    "            if i == 0:\n",
    "                exon_start = 0\n",
    "                exon_end = longest_transcript['Translation']['end'] - exon['start'] + 1\n",
    "            else:\n",
    "                exon_start = exon_end_last\n",
    "                exon_end = exon_start + exon['end'] - exon['start'] + 1\n",
    "        exon_end_last = exon_end\n",
    "        exon_feature = SeqFeature(FeatureLocation(exon_start, exon_end, strand=strand), type=\"exon\", id=exon['id'])\n",
    "        exon_feature.qualifiers['label'] = f\"{i+1}:{exon['id']}\"\n",
    "        exon_feature.qualifiers['ApEinfo_fwdcolor'] = random_color()\n",
    "        cds_record.features.append(exon_feature)\n",
    "        i += 1\n",
    "        if gene_id in roi:\n",
    "            if roi_start - 2 <= i <= roi_end + 2:\n",
    "                roi[gene_id]['ROI_flank_seq'] += str(cds_seq[exon_start:exon_end])\n",
    "            if roi_start <= i <= roi_end:\n",
    "                roi[gene_id]['ROI_seq'] += str(cds_seq[exon_start:exon_end])\n",
    "    if gene_id in roi:\n",
    "        roi[gene_id]['ROI_len'] = len(roi[gene_id]['ROI_seq'])\n",
    "        roi[gene_id]['ROI_flank_len'] = len(roi[gene_id]['ROI_flank_seq'])\n",
    "    if gene_id in primer_dict:\n",
    "        for primer in primer_dict[gene_id]:\n",
    "            matches = [match.start() for match in re.finditer(primer, str(cds_seq), re.IGNORECASE)]\n",
    "            if matches:\n",
    "                for match in matches:\n",
    "                    primer_feature = SeqFeature(FeatureLocation(match, match + len(primer)), type=\"primer\", id=primer)\n",
    "                    primer_feature.qualifiers['label'] = f\"primer{primer_dict[gene_id].index(primer)+1}:{primer}\"\n",
    "                    primer_feature.qualifiers['ApEinfo_fwdcolor'] = random_color()\n",
    "                    cds_record.features.append(primer_feature)\n",
    "            rev_primer = str(Seq(primer).reverse_complement())\n",
    "            rev_matches = [match.start() for match in re.finditer(rev_primer, str(cds_seq), re.IGNORECASE)]\n",
    "            if rev_matches:\n",
    "                for match in rev_matches:\n",
    "                    primer_feature = SeqFeature(FeatureLocation(match, match + len(rev_primer), strand=-1), type=\"primer\", id=rev_primer)\n",
    "                    primer_feature.qualifiers['label'] = f\"primer{primer_dict[gene_id].index(primer)+1}:{primer}\"\n",
    "                    primer_feature.qualifiers['ApEinfo_fwdcolor'] = random_color()\n",
    "                    cds_record.features.append(primer_feature)\n",
    "    output_folder = f\"./{gene_name}{gene_id}\"\n",
    "    output_path = os.path.join(output_folder, f\"{longest_transcript_id}.ape\")\n",
    "    if not os.path.exists(output_folder):\n",
    "        os.makedirs(output_folder)\n",
    "    with open(output_path, \"w\") as output_handle:\n",
    "        SeqIO.write(cds_record, output_handle, \"genbank\")\n",
    "\n",
    "gene_name_list = {}\n",
    "roi = {}\n",
    "for gene_id in gene_id_list:\n",
    "    annotate_gene(gene_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Designing Primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "design_internal_primer = True\n",
    "pcr_product_size_range = [[30, 1000]]\n",
    "genome = \"Astatotilapia_calliptera.fAstCal1.3.113\"\n",
    "email = \"your.email@example.com\"\n",
    "species = \" \".join(genome.split(\".\")[0].split(\"_\")[:2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ROI length: 215, using flanking sequence\n",
      "Primers for ENSACLG00000016414:\n",
      "Primer Pair 1:\n",
      "  Left Primer: AGGTGTTTGATGGCGACGAT\n",
      "  Right Primer: TGCTGTCTGAGACGGAAACC\n",
      "  Internal Primer: TCGCTGAAGGACGGAGGCGA\n",
      "L Primer BLAST Results:\n",
      "  Number of alignments: 15\n",
      "  Top hit: 2:10,450,246-10,478,221 ENSACLG00000016414  Astatotilapia calliptera diacylglycerol kinase theta (dgkq), transcript variant X1, mRNA, E-value: 0.00342527\n",
      "R Primer BLAST Results:\n",
      "  Number of alignments: 13\n",
      "  Top hit: 2:10,450,246-10,478,221 ENSACLG00000016414  Astatotilapia calliptera diacylglycerol kinase theta (dgkq), transcript variant X1, mRNA, E-value: 0.00342527\n",
      "I Primer BLAST Results:\n",
      "  Number of alignments: 25\n",
      "  Top hit: 2:10,450,246-10,478,221 ENSACLG00000016414  Astatotilapia calliptera diacylglycerol kinase theta (dgkq), transcript variant X1, mRNA, E-value: 0.00342526\n",
      "Common Hits for Primer Pair 1:\n",
      "  Number of common hits: 1\n",
      "  Hit 1: 2:10,450,246-10,478,221 ENSACLG00000016414  Astatotilapia calliptera diacylglycerol kinase theta (dgkq), transcript variant X1, mRNA, E-value: 0.00342527\n",
      "Primer Pair 2:\n",
      "  Left Primer: AGGTGTTTGATGGCGACGAT\n",
      "  Right Primer: ACCGAAATGAAAGCAACGCC\n",
      "  Internal Primer: TCGCTGAAGGACGGAGGCGA\n",
      "L Primer BLAST Results:\n",
      "  Number of alignments: 15\n",
      "  Top hit: 2:10,450,246-10,478,221 ENSACLG00000016414  Astatotilapia calliptera diacylglycerol kinase theta (dgkq), transcript variant X1, mRNA, E-value: 0.00342527\n",
      "R Primer BLAST Results:\n",
      "  Number of alignments: 11\n",
      "  Top hit: 2:10,450,246-10,478,221 ENSACLG00000016414  Astatotilapia calliptera diacylglycerol kinase theta (dgkq), transcript variant X2, mRNA, E-value: 6.19303\n",
      "I Primer BLAST Results:\n",
      "  Number of alignments: 25\n",
      "  Top hit: 2:10,450,246-10,478,221 ENSACLG00000016414  Astatotilapia calliptera diacylglycerol kinase theta (dgkq), transcript variant X1, mRNA, E-value: 0.00342526\n",
      "Common Hits for Primer Pair 2:\n",
      "  Number of common hits: 1\n",
      "  Hit 1: 2:10,450,246-10,478,221 ENSACLG00000016414  Astatotilapia calliptera diacylglycerol kinase theta (dgkq), transcript variant X1, mRNA, E-value: 0.00342527\n",
      "Primer Pair 3:\n",
      "  Left Primer: GGTGTTTGATGGCGACGATG\n",
      "  Right Primer: TGCTGTCTGAGACGGAAACC\n",
      "  Internal Primer: TCGCTGAAGGACGGAGGCGA\n",
      "L Primer BLAST Results:\n",
      "  Number of alignments: 12\n",
      "  Top hit: 2:10,450,246-10,478,221 ENSACLG00000016414  Astatotilapia calliptera diacylglycerol kinase theta (dgkq), transcript variant X1, mRNA, E-value: 0.00342527\n",
      "R Primer BLAST Results:\n",
      "  Number of alignments: 13\n",
      "  Top hit: 2:10,450,246-10,478,221 ENSACLG00000016414  Astatotilapia calliptera diacylglycerol kinase theta (dgkq), transcript variant X1, mRNA, E-value: 0.00342527\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[41], line 176\u001b[0m\n\u001b[1;32m    174\u001b[0m result_file \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m./\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mpath\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m/Primer\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mi\u001b[38;5;241m+\u001b[39m\u001b[38;5;241m1\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;132;01m{\u001b[39;00mprimer_type\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m_blast.xml\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    175\u001b[0m blastn(\u001b[38;5;28mstr\u001b[39m(primer_seq), result_file)\n\u001b[0;32m--> 176\u001b[0m blast_df \u001b[38;5;241m=\u001b[39m \u001b[43mparse_blast_xml\u001b[49m\u001b[43m(\u001b[49m\u001b[43mresult_file\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    177\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m idx, row \u001b[38;5;129;01min\u001b[39;00m blast_df\u001b[38;5;241m.\u001b[39miterrows():\n\u001b[1;32m    178\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m row[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mID\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m==\u001b[39m gene_id:\n",
      "Cell \u001b[0;32mIn[41], line 62\u001b[0m, in \u001b[0;36mparse_blast_xml\u001b[0;34m(xml_file)\u001b[0m\n\u001b[1;32m     60\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(xml_file) \u001b[38;5;28;01mas\u001b[39;00m handle:\n\u001b[1;32m     61\u001b[0m     blast_record \u001b[38;5;241m=\u001b[39m NCBIXML\u001b[38;5;241m.\u001b[39mread(handle)\n\u001b[0;32m---> 62\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mhandle_blast_record\u001b[49m\u001b[43m(\u001b[49m\u001b[43mblast_record\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[41], line 39\u001b[0m, in \u001b[0;36mhandle_blast_record\u001b[0;34m(blast_record)\u001b[0m\n\u001b[1;32m     37\u001b[0m         ncbi_id \u001b[38;5;241m=\u001b[39m alignment\u001b[38;5;241m.\u001b[39mtitle\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m|\u001b[39m\u001b[38;5;124m'\u001b[39m)[\u001b[38;5;241m3\u001b[39m]\n\u001b[1;32m     38\u001b[0m         gene_id \u001b[38;5;241m=\u001b[39m get_ensembl_id_from_ncbi_id(ncbi_id)\n\u001b[0;32m---> 39\u001b[0m         position \u001b[38;5;241m=\u001b[39m \u001b[43mget_absolute_position\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgene_id\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mif\u001b[39;00m gene_id \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     40\u001b[0m         results\u001b[38;5;241m.\u001b[39mappend({\n\u001b[1;32m     41\u001b[0m             \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mID\u001b[39m\u001b[38;5;124m'\u001b[39m: gene_id \u001b[38;5;28;01mif\u001b[39;00m gene_id \u001b[38;5;28;01melse\u001b[39;00m alignment\u001b[38;5;241m.\u001b[39mtitle\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m|\u001b[39m\u001b[38;5;124m'\u001b[39m)[\u001b[38;5;241m1\u001b[39m],\n\u001b[1;32m     42\u001b[0m             \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mPosition\u001b[39m\u001b[38;5;124m'\u001b[39m: position \u001b[38;5;28;01mif\u001b[39;00m position \u001b[38;5;28;01melse\u001b[39;00m alignment\u001b[38;5;241m.\u001b[39mtitle\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m|\u001b[39m\u001b[38;5;124m'\u001b[39m)[\u001b[38;5;241m3\u001b[39m],\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     55\u001b[0m             \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSubject Sequence\u001b[39m\u001b[38;5;124m'\u001b[39m: hsp\u001b[38;5;241m.\u001b[39msbjct\n\u001b[1;32m     56\u001b[0m         })\n\u001b[1;32m     57\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m pd\u001b[38;5;241m.\u001b[39mDataFrame(results)\n",
      "Cell \u001b[0;32mIn[41], line 22\u001b[0m, in \u001b[0;36mget_absolute_position\u001b[0;34m(ensembl_id)\u001b[0m\n\u001b[1;32m     20\u001b[0m url \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhttps://rest.ensembl.org/lookup/id/\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mensembl_id\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m?expand=1\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     21\u001b[0m headers \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mContent-Type\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mapplication/json\u001b[39m\u001b[38;5;124m\"\u001b[39m}\n\u001b[0;32m---> 22\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[43mrequests\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     23\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m response\u001b[38;5;241m.\u001b[39mstatus_code \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m200\u001b[39m:\n\u001b[1;32m     24\u001b[0m     data \u001b[38;5;241m=\u001b[39m response\u001b[38;5;241m.\u001b[39mjson()\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/seq/lib/python3.12/site-packages/requests/api.py:73\u001b[0m, in \u001b[0;36mget\u001b[0;34m(url, params, **kwargs)\u001b[0m\n\u001b[1;32m     62\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget\u001b[39m(url, params\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m     63\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"Sends a GET request.\u001b[39;00m\n\u001b[1;32m     64\u001b[0m \n\u001b[1;32m     65\u001b[0m \u001b[38;5;124;03m    :param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     70\u001b[0m \u001b[38;5;124;03m    :rtype: requests.Response\u001b[39;00m\n\u001b[1;32m     71\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 73\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mget\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparams\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/seq/lib/python3.12/site-packages/requests/api.py:59\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[38;5;66;03m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[1;32m     56\u001b[0m \u001b[38;5;66;03m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;66;03m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m sessions\u001b[38;5;241m.\u001b[39mSession() \u001b[38;5;28;01mas\u001b[39;00m session:\n\u001b[0;32m---> 59\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43msession\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/seq/lib/python3.12/site-packages/requests/sessions.py:589\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    584\u001b[0m send_kwargs \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    585\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtimeout\u001b[39m\u001b[38;5;124m\"\u001b[39m: timeout,\n\u001b[1;32m    586\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m\"\u001b[39m: allow_redirects,\n\u001b[1;32m    587\u001b[0m }\n\u001b[1;32m    588\u001b[0m send_kwargs\u001b[38;5;241m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 589\u001b[0m resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprep\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43msend_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    591\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/seq/lib/python3.12/site-packages/requests/sessions.py:703\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    700\u001b[0m start \u001b[38;5;241m=\u001b[39m preferred_clock()\n\u001b[1;32m    702\u001b[0m \u001b[38;5;66;03m# Send the request\u001b[39;00m\n\u001b[0;32m--> 703\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43madapter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    705\u001b[0m \u001b[38;5;66;03m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    706\u001b[0m elapsed \u001b[38;5;241m=\u001b[39m preferred_clock() \u001b[38;5;241m-\u001b[39m start\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/seq/lib/python3.12/site-packages/requests/adapters.py:667\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    664\u001b[0m     timeout \u001b[38;5;241m=\u001b[39m TimeoutSauce(connect\u001b[38;5;241m=\u001b[39mtimeout, read\u001b[38;5;241m=\u001b[39mtimeout)\n\u001b[1;32m    666\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 667\u001b[0m     resp \u001b[38;5;241m=\u001b[39m \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43murlopen\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    668\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    669\u001b[0m \u001b[43m        \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    670\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    671\u001b[0m \u001b[43m        \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    672\u001b[0m \u001b[43m        \u001b[49m\u001b[43mredirect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    673\u001b[0m \u001b[43m        \u001b[49m\u001b[43massert_same_host\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    674\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    675\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    676\u001b[0m \u001b[43m        \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmax_retries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    677\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    678\u001b[0m \u001b[43m        \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    679\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    681\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (ProtocolError, \u001b[38;5;167;01mOSError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[1;32m    682\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m(err, request\u001b[38;5;241m=\u001b[39mrequest)\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/seq/lib/python3.12/site-packages/urllib3/connectionpool.py:789\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, preload_content, decode_content, **response_kw)\u001b[0m\n\u001b[1;32m    786\u001b[0m response_conn \u001b[38;5;241m=\u001b[39m conn \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m release_conn \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    788\u001b[0m \u001b[38;5;66;03m# Make the request on the HTTPConnection object\u001b[39;00m\n\u001b[0;32m--> 789\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    790\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    791\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    792\u001b[0m \u001b[43m    \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    793\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout_obj\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    794\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    795\u001b[0m \u001b[43m    \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    796\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    797\u001b[0m \u001b[43m    \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mretries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    798\u001b[0m \u001b[43m    \u001b[49m\u001b[43mresponse_conn\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mresponse_conn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    799\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpreload_content\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    800\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdecode_content\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    801\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mresponse_kw\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    802\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    804\u001b[0m \u001b[38;5;66;03m# Everything went great!\u001b[39;00m\n\u001b[1;32m    805\u001b[0m clean_exit \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/seq/lib/python3.12/site-packages/urllib3/connectionpool.py:536\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, body, headers, retries, timeout, chunked, response_conn, preload_content, decode_content, enforce_content_length)\u001b[0m\n\u001b[1;32m    534\u001b[0m \u001b[38;5;66;03m# Receive the response from the server\u001b[39;00m\n\u001b[1;32m    535\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 536\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetresponse\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    537\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (BaseSSLError, \u001b[38;5;167;01mOSError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    538\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_raise_timeout(err\u001b[38;5;241m=\u001b[39me, url\u001b[38;5;241m=\u001b[39murl, timeout_value\u001b[38;5;241m=\u001b[39mread_timeout)\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/seq/lib/python3.12/site-packages/urllib3/connection.py:507\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    504\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mresponse\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m HTTPResponse\n\u001b[1;32m    506\u001b[0m \u001b[38;5;66;03m# Get the response from http.client.HTTPConnection\u001b[39;00m\n\u001b[0;32m--> 507\u001b[0m httplib_response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetresponse\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    509\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    510\u001b[0m     assert_header_parsing(httplib_response\u001b[38;5;241m.\u001b[39mmsg)\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/seq/lib/python3.12/http/client.py:1428\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1426\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1427\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1428\u001b[0m         \u001b[43mresponse\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbegin\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1429\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m:\n\u001b[1;32m   1430\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclose()\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/seq/lib/python3.12/http/client.py:331\u001b[0m, in \u001b[0;36mHTTPResponse.begin\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    329\u001b[0m \u001b[38;5;66;03m# read until we get a non-100 response\u001b[39;00m\n\u001b[1;32m    330\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[0;32m--> 331\u001b[0m     version, status, reason \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_read_status\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    332\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m status \u001b[38;5;241m!=\u001b[39m CONTINUE:\n\u001b[1;32m    333\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/seq/lib/python3.12/http/client.py:292\u001b[0m, in \u001b[0;36mHTTPResponse._read_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    291\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_read_status\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m--> 292\u001b[0m     line \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m(\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mreadline\u001b[49m\u001b[43m(\u001b[49m\u001b[43m_MAXLINE\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124miso-8859-1\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    293\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(line) \u001b[38;5;241m>\u001b[39m _MAXLINE:\n\u001b[1;32m    294\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m LineTooLong(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstatus line\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/seq/lib/python3.12/socket.py:720\u001b[0m, in \u001b[0;36mSocketIO.readinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    718\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[1;32m    719\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 720\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sock\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrecv_into\u001b[49m\u001b[43m(\u001b[49m\u001b[43mb\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    721\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m timeout:\n\u001b[1;32m    722\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_timeout_occurred \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/seq/lib/python3.12/ssl.py:1251\u001b[0m, in \u001b[0;36mSSLSocket.recv_into\u001b[0;34m(self, buffer, nbytes, flags)\u001b[0m\n\u001b[1;32m   1247\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m flags \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m   1248\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   1249\u001b[0m           \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnon-zero flags not allowed in calls to recv_into() on \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m\n\u001b[1;32m   1250\u001b[0m           \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m)\n\u001b[0;32m-> 1251\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnbytes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbuffer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1252\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1253\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28msuper\u001b[39m()\u001b[38;5;241m.\u001b[39mrecv_into(buffer, nbytes, flags)\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/seq/lib/python3.12/ssl.py:1103\u001b[0m, in \u001b[0;36mSSLSocket.read\u001b[0;34m(self, len, buffer)\u001b[0m\n\u001b[1;32m   1101\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1102\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m buffer \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 1103\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sslobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbuffer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1104\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1105\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sslobj\u001b[38;5;241m.\u001b[39mread(\u001b[38;5;28mlen\u001b[39m)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "annotaion_file = f\"./genome/{genome}.gtf\"\n",
    "genome_file = f\"./genome/{genome.rsplit('.', 1)[0]}.dna_sm.toplevel.fa\"\n",
    "blast_db = f\"./blastdb/{genome}\"\n",
    "\n",
    "import primer3\n",
    "import pandas as pd\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Blast import NCBIWWW\n",
    "from Bio import Entrez\n",
    "Entrez.email = email\n",
    "\n",
    "def get_ensembl_id_from_ncbi_id(ncbi_id):\n",
    "    url = f\"https://mygene.info/v3/query?q={ncbi_id}&fields=ensembl.gene\"\n",
    "    response = requests.get(url)\n",
    "    if response.status_code == 200:\n",
    "        data = response.json()\n",
    "        if 'hits' in data and len(data['hits']) > 0:\n",
    "            ensembl_id = data['hits'][0].get('ensembl', {}).get('gene')\n",
    "            return ensembl_id\n",
    "    return None\n",
    "\n",
    "def get_absolute_position(ensembl_id):\n",
    "    url = f\"https://rest.ensembl.org/lookup/id/{ensembl_id}?expand=1\"\n",
    "    headers = {\"Content-Type\": \"application/json\"}\n",
    "    response = requests.get(url, headers=headers)\n",
    "    if response.status_code == 200:\n",
    "        data = response.json()\n",
    "        if 'seq_region_name' in data and 'start' in data and 'end' in data:\n",
    "            chrom = data['seq_region_name']\n",
    "            start = data['start']\n",
    "            end = data['end']\n",
    "            absolute_position = f\"{chrom}:{start:,}-{end:,}\"\n",
    "            return absolute_position\n",
    "    return None\n",
    "\n",
    "def handle_blast_record(blast_record):\n",
    "    results = []\n",
    "    for alignment in blast_record.alignments:\n",
    "        for hsp in alignment.hsps:\n",
    "            ncbi_id = alignment.title.split('|')[3]\n",
    "            gene_id = get_ensembl_id_from_ncbi_id(ncbi_id)\n",
    "            position = get_absolute_position(gene_id) if gene_id else \"\"\n",
    "            results.append({\n",
    "                'ID': gene_id if gene_id else alignment.title.split('|')[1],\n",
    "                'Position': position if position else alignment.title.split('|')[3],\n",
    "                'Title': alignment.title.split(':')[1],\n",
    "                'Length': alignment.length,\n",
    "                'E-value': hsp.expect,\n",
    "                'Score': hsp.score,\n",
    "                'Query Start': hsp.query_start,\n",
    "                'Query End': hsp.query_end,\n",
    "                'Subject Start': hsp.sbjct_start,\n",
    "                'Subject End': hsp.sbjct_end,\n",
    "                'Identities': hsp.identities,\n",
    "                'Gaps': hsp.gaps,\n",
    "                'Positives': hsp.positives,\n",
    "                'Query Sequence': hsp.query,\n",
    "                'Subject Sequence': hsp.sbjct\n",
    "            })\n",
    "    return pd.DataFrame(results)\n",
    "\n",
    "def parse_blast_xml(xml_file):\n",
    "    with open(xml_file) as handle:\n",
    "        blast_record = NCBIXML.read(handle)\n",
    "    return handle_blast_record(blast_record)\n",
    "\n",
    "def blastn(sequence, output_file):\n",
    "    result_handle = NCBIWWW.qblast(\"blastn\", \"nt\", sequence, entrez_query=f\"{species}[ORGN]\")\n",
    "    with open(output_file, \"w\") as out_handle:\n",
    "        out_handle.write(result_handle.read())\n",
    "    result_handle.close()\n",
    "\n",
    "def get_primers(data):\n",
    "    roi_len = data['ROI_len']\n",
    "    roi_flank_seq = data['ROI_flank_seq']\n",
    "    roi_seq = data['ROI_seq']\n",
    "    exclude_roi = False\n",
    "    if any(lower <= roi_len <= upper for lower, upper in pcr_product_size_range):\n",
    "        sequence = roi_flank_seq\n",
    "        exclude_roi = True\n",
    "        start_pos = roi_flank_seq.find(roi_seq)\n",
    "        print(f\"ROI length: {roi_len}, using flanking sequence\")\n",
    "    else:\n",
    "        sequence = roi_seq\n",
    "        print(f\"ROI length: {roi_len}, using ROI sequence\")\n",
    "    seq_args = {\n",
    "        'SEQUENCE_TEMPLATE': sequence,\n",
    "    }\n",
    "    global_args = {\n",
    "        'SEQUENCE_EXCLUDED_REGION': [[start_pos, roi_len]] if exclude_roi else [],\n",
    "        'SEQUENCE_INTERNAL_EXCLUDED_REGION': [[1, start_pos - 1], [start_pos + roi_len, len(sequence) - start_pos - roi_len]] if exclude_roi else [],\n",
    "        'PRIMER_PICK_INTERNAL_OLIGO': 1 if design_internal_primer else 0,\n",
    "        'PRIMER_OPT_SIZE': 20,\n",
    "        'PRIMER_INTERNAL_MAX_SELF_END': 8,\n",
    "        'PRIMER_MIN_SIZE': 18,\n",
    "        'PRIMER_MAX_SIZE': 27,\n",
    "        'PRIMER_OPT_TM': 60.0,\n",
    "        'PRIMER_MIN_TM': 57.0,\n",
    "        'PRIMER_MAX_TM': 63.0,\n",
    "        'PRIMER_OPT_GC_PERCENT': 50.0,\n",
    "        'PRIMER_MIN_GC': 20.0,\n",
    "        'PRIMER_MAX_GC': 80.0,\n",
    "        'PRIMER_MAX_POLY_X': 5,\n",
    "        'PRIMER_INTERNAL_MAX_POLY_X': 5,\n",
    "        'PRIMER_SALT_MONOVALENT': 50.0,\n",
    "        'PRIMER_SALT_DIVALENT': 1.5,\n",
    "        'PRIMER_DNA_CONC': 50.0,\n",
    "        'PRIMER_MAX_NS_ACCEPTED': 0,\n",
    "        'PRIMER_MAX_SELF_ANY': 8,\n",
    "        'PRIMER_MAX_SELF_END': 3,\n",
    "        'PRIMER_PAIR_MAX_COMPL_ANY': 8,\n",
    "        'PRIMER_PAIR_MAX_COMPL_END': 3,\n",
    "        'PRIMER_PRODUCT_SIZE_RANGE': pcr_product_size_range,\n",
    "        'PRIMER_NUM_RETURN': 5\n",
    "    }\n",
    "    primers = primer3.bindings.design_primers(seq_args, global_args)\n",
    "    return primers\n",
    "\n",
    "for gene_id, data in roi.items():\n",
    "    records = []\n",
    "    path = f\"./{gene_name_list[gene_id]}{gene_id}\"\n",
    "    primers = get_primers(data)\n",
    "    with open(f\"{path}/Primer_summary.txt\", \"a\") as summary_file:\n",
    "        summary_file.write(f\"Primers for {gene_id}:\")\n",
    "        print(f\"Primers for {gene_id}:\")\n",
    "        for i in range(primers['PRIMER_LEFT_NUM_RETURNED']):\n",
    "            left_primer_seq = Seq(primers[f'PRIMER_LEFT_{i}_SEQUENCE'])\n",
    "            right_primer_seq = Seq(primers[f'PRIMER_RIGHT_{i}_SEQUENCE'])\n",
    "            if design_internal_primer: internal_primer_seq = Seq(primers[f'PRIMER_INTERNAL_{i}_SEQUENCE'])\n",
    "            summary_file.write(f\"Primer Pair {i+1}:\\n\")\n",
    "            summary_file.write(f\"  Left Primer: {left_primer_seq}\\n\")\n",
    "            summary_file.write(f\"  Right Primer: {right_primer_seq}\\n\")\n",
    "            if design_internal_primer: summary_file.write(f\"  Internal Primer: {internal_primer_seq}\\n\")\n",
    "            print(f\"Primer Pair {i+1}:\")\n",
    "            print(f\"  Left Primer: {left_primer_seq}\")\n",
    "            print(f\"  Right Primer: {right_primer_seq}\")\n",
    "            if design_internal_primer: print(f\"  Internal Primer: {internal_primer_seq}\")\n",
    "            records.append(SeqRecord(left_primer_seq, id=f\"{i+1}L\", description=\"\"))\n",
    "            records.append(SeqRecord(right_primer_seq, id=f\"{i+1}R\", description=\"\"))\n",
    "            if design_internal_primer:\n",
    "                records.append(SeqRecord(internal_primer_seq, id=f\"{i+1}I\", description=\"\"))\n",
    "                array = [('L', left_primer_seq), ('R', right_primer_seq), ('I', internal_primer_seq)]\n",
    "            else:\n",
    "                array = [('L', left_primer_seq), ('R', right_primer_seq)] \n",
    "            for primer_type, primer_seq in array:\n",
    "                result_file = f\"./{path}/Primer{i+1}{primer_type}_blast.xml\"\n",
    "                blastn(str(primer_seq), result_file)\n",
    "                blast_df = parse_blast_xml(result_file)\n",
    "                for idx, row in blast_df.iterrows():\n",
    "                    if row['ID'] == gene_id:\n",
    "                        blast_df = blast_df.drop(idx)\n",
    "                if not blast_df.empty:\n",
    "                    summary_file.write(f\"{primer_type} Primer BLAST Results:\\n\")\n",
    "                    summary_file.write(f\"  Number of alignments: {len(blast_df)}\\n\")\n",
    "                    summary_file.write(f\"  Top hit: {blast_df.iloc[0]['Position']} {blast_df.iloc[0]['ID']} {blast_df.iloc[0]['Title']}, E-value: {blast_df.iloc[0]['E-value']}\\n\")\n",
    "                    print(f\"{primer_type} Primer BLAST Results:\")\n",
    "                    print(f\"  Number of alignments: {len(blast_df)}\")\n",
    "                    print(f\"  Top hit: {blast_df.iloc[0]['Position']} {blast_df.iloc[0]['ID']} {blast_df.iloc[0]['Title']}, E-value: {blast_df.iloc[0]['E-value']}\")\n",
    "                else:\n",
    "                    summary_file.write(f\"{primer_type} Primer BLAST Results: No hits found\\n\")\n",
    "                    print(f\"{primer_type} Primer BLAST Results: No hits found\")\n",
    "                blast_df.to_csv(f\"./{path}/Primer{i+1}{primer_type}_blast.csv\", index=False)\n",
    "                left_blast_df = pd.read_csv(f\"./{path}/Primer{i+1}L_blast.csv\")\n",
    "                right_blast_df = pd.read_csv(f\"./{path}/Primer{i+1}R_blast.csv\")\n",
    "                common_hits = pd.merge(left_blast_df, right_blast_df, on='ID', suffixes=('', '_right'), how='outer')\n",
    "                if design_internal_primer:\n",
    "                    internal_blast_df = pd.read_csv(f\"./{path}/Primer{i+1}I_blast.csv\")\n",
    "                    common_hits = pd.merge(common_hits, internal_blast_df, on='ID', suffixes=('', '_internal'))\n",
    "                    common_hits = common_hits.drop_duplicates(subset=['ID']).reset_index(drop=True)\n",
    "                summary_file.write(f\"Common Hits for Primer Pair {i+1}:\\n\")\n",
    "                summary_file.write(f\"  Number of common hits: {len(common_hits)}\\n\")\n",
    "                if not common_hits.empty:\n",
    "                    for index, row in common_hits.iterrows():\n",
    "                        summary_file.write(f\"  Hit {index+1}: {row['Position']} {row['ID']} {row['Title']}, E-value: {row['E-value']}\\n\")\n",
    "                print(f\"Common Hits for Primer Pair {i+1}:\")\n",
    "                print(f\"  Number of common hits: {len(common_hits)}\")\n",
    "                if not common_hits.empty:\n",
    "                    for index, row in common_hits.iterrows():\n",
    "                        print(f\"  Hit {index+1}: {row['Position']} {row['ID']} {row['Title']}, E-value: {row['E-value']}\")\n",
    "                common_hits.to_csv(f\"./{path}/Primer{i+1}_common_hits.csv\", index=False)\n",
    "    SeqIO.write(records, f\"{path}/Primers.fa\", \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Backup Scripts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ENSACLG00000004087 E002\n",
    "ENSACLG00000026087 E008?\n",
    "ENSACLG00000012728 E001?\n",
    "ENSACLG00000020646 E004\n",
    "ENSACLG00000011505 E012\n",
    "ENSACLG00000003876 E042\n",
    "ENSACLG00000025257 E012/\n",
    "ENSACLG00000002025 E014/\n",
    "**ENSACLG00000016414 E015**\n",
    "**ENSACLG00000015604 E005**\n",
    "**ENSACLG00000010182 E020**\n",
    "**ENSACLG00000018072 E002**\n",
    "**ENSACLG00000021184 E002**\n",
    "**ENSACLG00000027919 E018**\n",
    "**ENSACLG00000016414 E017**\n",
    "**ENSACLG00000003858 E006**\n",
    "**ENSACLG00000022676 E004**\n",
    "**ENSACLG00000015604 E002**\n",
    "**ENSACLG00000018499 E005**\n",
    "**ENSACLG00000009499 E001**\n",
    "**ENSACLG00000015133 E018**\n",
    "**ENSACLG00000010075 E011**\n",
    "ENSACLG00000000969\n",
    "ENSACLG00000004086\n",
    "ENSACLG00000019020\n",
    "ENSACLG00000003013\n",
    "**ENSACLG00000024008**\n",
    "ENSACLG00000009059\n",
    "**ENSACLG00000002869**\n",
    "ENSACLG00000021745\n",
    "ENSACLG00000027444\\*\n",
    "ENSACLG00000008688\n",
    "ENSACLG00000004444\n",
    "ENSACLG00000015800 TP\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "import subprocess\n",
    "\n",
    "packages_to_install = ['numpy', 'pandas', 'pingouin', 'patchworklib', 'plotnine', 'scikit-posthocs', 'statsmodels']\n",
    "env_name = 'biopython'\n",
    "\n",
    "for package in packages_to_install:\n",
    "    subprocess.run(['conda', 'install', '-n', env_name, package, '-y'])\n",
    "\"\"\" \n",
    "# gtf_columns = [\n",
    "#     'seqname', 'source', 'feature', 'start', 'end',\n",
    "#     'score', 'strand', 'frame', 'attribute'\n",
    "# ]\n",
    "# gtf_df = pd.read_csv(\n",
    "#     annotaion_file, \n",
    "#     sep='\\t', \n",
    "#     comment='#', \n",
    "#     header=None, \n",
    "#     names=gtf_columns\n",
    "# )\n",
    "# genes_df = gtf_df[gtf_df['feature'] == 'gene']\n",
    "# #import re\n",
    "# def parse_attributes(attribute_string):\n",
    "#     attrs = {}\n",
    "#     for attr in attribute_string.split(';'):\n",
    "#         if attr.strip() == '':\n",
    "#             continue\n",
    "#         key, value = attr.strip().split(' ', 1)\n",
    "#         attrs[key] = value.replace('\"', '')\n",
    "#     return attrs\n",
    "# attributes = genes_df['attribute'].apply(parse_attributes)\n",
    "# attributes_df = pd.json_normalize(attributes)\n",
    "# genes_df = genes_df.join(attributes_df)\n",
    "# print(genes_df.head())\n",
    "# blast_df = parse_blast_xml(\"blast_results.xml\")\n",
    "# print(blast_df.head())\n",
    "# # Merge BLAST results with gene annotations\n",
    "# # Assuming 'Hit ID' corresponds to gene IDs in the GTF file, adjust as necessary\n",
    "# merged_df = blast_df.merge(\n",
    "#     genes_df, \n",
    "#     left_on='Hit ID', \n",
    "#     right_on='gene_id', \n",
    "#     how='left'\n",
    "# )\n",
    "# print(merged_df.head())\n",
    "# Save merged results to CSV for user-friendly viewing\n",
    "# merged_df.to_csv(\"merged_blast_annotations.csv\", index=False)\n",
    "\n",
    "# with open(output_file) as result_handle:\n",
    "#     blast_record = NCBIXML.read(result_handle)\n",
    "# for alignment in blast_record.alignments:\n",
    "#     for hsp in alignment.hsps:\n",
    "#         print(f\"****Alignment****\")\n",
    "#         print(f\"Sequence: {alignment.title}\")\n",
    "#         print(f\"Length: {alignment.length}\")\n",
    "#         print(f\"e-value: {hsp.expect}\")\n",
    "#         print(hsp.query)\n",
    "#         print(hsp.match)\n",
    "#         print(hsp.sbjct)\n",
    "    # if not enable_online_blast:\n",
    "    #     blastn(f\"{path}/Primers.fa\", f\"{path}/Primers_blast.xml\")\n",
    "    #     with open(f\"{path}/Primers_blast.xml\") as result_handle:\n",
    "    #         blast_records = NCBIXML.parse(result_handle)\n",
    "    #         for blast_record in blast_records:\n",
    "    #             print(f\"Query: {blast_record.query}\")\n",
    "# if enable_online_blast:\n",
    "#     from Bio import Entrez\n",
    "#     from Bio.Blast import NCBIWWW\n",
    "#     Entrez.email = email\n",
    "#     def blastn(sequence, output_file):\n",
    "#         result_handle = NCBIWWW.qblast(\"blastn\", \"nt\", sequence, entrez_query=f\"{species}[ORGN]\")\n",
    "#         with open(output_file, \"w\") as out_handle:\n",
    "#             out_handle.write(result_handle.read())\n",
    "#         result_handle.close()\n",
    "# else:\n",
    "#     if not os.path.exists(os.path.dirname(blast_db)):\n",
    "#         os.makedirs(os.path.dirname(blast_db))\n",
    "#     if not any(fname.endswith(\".ndb\") for fname in os.listdir(os.path.dirname(blast_db))):\n",
    "#         import urllib.request\n",
    "#         import re\n",
    "#         blast_db_url = \"https://ftp.ensembl.org/pub/current_ncbi_blast/genomic/\"\n",
    "#         file_pattern = f\"{species.replace(' ', '_')}.*.dna_sm.toplevel.fa.*\"\n",
    "#         file_pattern = re.compile(file_pattern.replace(\".\", r\"\\.\").replace(\"*\", \".*\").replace(\"?\", \".\"))\n",
    "#         with urllib.request.urlopen(blast_db_url) as response:\n",
    "#             html = response.read().decode('utf-8')\n",
    "#             file_list = re.findall(r'href=\"([^\"]+)\"', html)\n",
    "#         for file in file_list:\n",
    "#             if file_pattern.match(file):\n",
    "#                 file_url = f\"{blast_db_url}{file}\"\n",
    "#                 print(f\"Downloading {file_url}\")\n",
    "#                 local_file = os.path.join(os.path.dirname(blast_db), file)\n",
    "#                 blast_db = os.path.splitext(local_file)[0]\n",
    "#                 urllib.request.urlretrieve(file_url, local_file)\n",
    "#         if not any(fname.endswith(\".ndb\") for fname in os.listdir(os.path.dirname(blast_db))):\n",
    "#             if not os.path.exists(blast_db + \".asnb\"):\n",
    "#                 subprocess.run(['convert2blastmask', '-in', genome_file, '-parse_seqids', '-masking_algorithm', 'repeat', '-masking_options', 'repeatmasker', '-outfmt', 'maskinfo_asn1_bin', '-out', f'{blast_db}.asnb'], check=True)\n",
    "#             subprocess.run(['makeblastdb', '-in', genome_file, '-dbtype', 'nucl', '-mask_data', f'{blast_db}.asnb', '-parse_seqids', '-title', species, '-out', blast_db], check=True)\n",
    "#     blast_db = os.path.splitext(next(fname for fname in os.listdir(os.path.dirname(blast_db)) if fname.endswith(\".ndb\")))[0]\n",
    "#     def blastn(blast_query, output_file):\n",
    "#         subprocess.run(['blastn', '-query', blast_query, '-db', blast_db, '-out', output_file, '-outfmt', '5', '-evalue', '1e-6'], check=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import webbrowser\n",
    "# df_results_group_counts = df_results_cleaned[df_csv_groupby].value_counts().sort_values(ascending=False)\n",
    "# df_results_group_max = df_results_group_counts[df_results_group_counts == df_results_group_counts.max()].index.tolist()\n",
    "# df_results_group_counts.to_csv(\"df_results_cleaned_onlyID.csv\", header=True)\n",
    "# for gene_id_list in df_results_group_max:\n",
    "#     webbrowser.open_new(\"https://www.ensembl.org/id/\" + gene_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bulk RNA Seq Analysis\n",
    "Run `./Salmon` first"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "seq",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
